File: android/app/src/main/java/pro/sixvpn/app/core/Base64Text.kt
package pro.sixvpn.app.core

/**
 * Base64, by hand, for the places this app meets it: a subscription blob and
 * the credential inside a `ss://` line.
 *
 * Written out rather than taken from a library because of where it has to run.
 * `java.util.Base64` is API 26 and this app runs from 24; `android.util.Base64`
 * exists on every version and cannot be called off a device, which would put
 * the parsing that decides what the tunnel connects to beyond the reach of a
 * test. Neither trade is worth it for twenty lines.
 *
 * Every form these strings arrive in is accepted — standard or URL-safe
 * alphabet, padded or not — because the subscription endpoints of the world
 * disagree about all three and the person pasting one has no idea which they
 * were handed.
 */
internal object Base64Text {

    private const val ALPHABET =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"

    /** The decoded text, or null when [text] is not base64 at all. */
    fun decode(text: String): String? {
        val normalised = text.replace('-', '+').replace('_', '/').trimEnd('=')
        if (normalised.isEmpty()) return null
        if (normalised.any { it !in ALPHABET }) return null

        val bytes = ArrayList<Byte>(normalised.length * 3 / 4 + 3)
        var buffer = 0
        var bits = 0
        for (character in normalised) {
            buffer = (buffer shl 6) or ALPHABET.indexOf(character)
            bits += 6
            if (bits >= 8) {
                bits -= 8
                bytes.add(((buffer shr bits) and 0xFF).toByte())
            }
        }
        if (bytes.isEmpty()) return null
        return String(bytes.toByteArray(), Charsets.UTF_8)
    }
}

File: ios/SyxVPNCore/Sources/SyxVPNCore/Base64Text.swift
import Foundation

/// Base64, by hand, for the places this app meets it: a subscription blob and
/// the credential inside an `ss://` line.
///
/// Written out rather than taken from Foundation because of what arrives here.
/// `Data(base64Encoded:)` refuses a string whose padding is missing and refuses
/// the URL-safe alphabet outright, and the subscription endpoints of the world
/// disagree about both — the person pasting one has no idea which they were
/// handed. Normalising into Foundation would be most of this file anyway.
enum Base64Text {

    private static let alphabet = Array(
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
    )

    private static let index: [Character: Int] = {
        var map: [Character: Int] = [:]
        for (position, character) in alphabet.enumerated() { map[character] = position }
        return map
    }()

    /// The decoded text, or nil when `text` is not base64 at all.
    static func decode(_ text: String) -> String? {
        var normalised = text.replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        while normalised.hasSuffix("=") { normalised.removeLast() }
        if normalised.isEmpty { return nil }

        var bytes: [UInt8] = []
        bytes.reserveCapacity(normalised.count * 3 / 4 + 3)
        var buffer = 0
        var bits = 0
        for character in normalised {
            guard let value = index[character] else { return nil }
            buffer = (buffer << 6) | value
            bits += 6
            if bits >= 8 {
                bits -= 8
                bytes.append(UInt8((buffer >> bits) & 0xFF))
            }
        }
        if bytes.isEmpty { return nil }
        return String(decoding: bytes, as: UTF8.self)
    }
}
